### ISSUE TITLE
Switching networks

### ISSUE THREAD

**Issue by forsomefun** [jrudess/mfc_cb.node#4]
Does not handle switching of networks at all.  TCP error
other node recording software does.

**Comment by jrudess** [comment]
I've seen the issue occur.  The difference is likely due to the switch to the mfcauto js library.

**Comment by jrudess** [comment]
I have put some initial fixes for this issue into the development branch.   Can you download that branch and see if the issues are resolved for you now?

The process seems to stay up reliably now when network connections go up and down, and model captures resume when network is back up, but a lot more testing is needed to know if it is truly stable.

**Comment by snickerpop** [comment]
IS it only ment to work with chaturbate now?. I couldn't get it to check for MFC models

**Comment by jrudess** [comment]
I'm not able to replicate that.  MFC seems to be working fine for me.  Is enableMFC set to true in config.yml?

**Comment by snickerpop** [comment]
I am not sure what happened.  I redownloaded everything and was able to 
I was able to fix it through

**Comment by snickerpop** [comment]
Comcast limits data to 1TB per month. Streaming models takes up so much bandwidth. 
Anyways I have my Computer setup were the wifi is connected, but I usually disable the ethernet save for a few hours at night.

I reenabled the Ethernet for 10 minutes and then disable it again. During this time everything seems to record fine. 

I also did it for another 1hr and when I disable it again the recording went on normal. 
I haven't gotten any error messages now.

**Comment by jrudess** [comment]
Thanks for the feedback.  I will run with the development version for a week or so, then push it out to a release.

**Comment by snickerpop** [comment]
I did a few more test to see how it affected files. As a based my previous feedback on how the cmder tool responded to the changes. 
You can see a small jump in the recording when the network switches, but to me, that sounds more like an issue with the hardware and a windows delay in detecting a change in networks. 
However after that things will record fine, and it doesn;t force creation of a new file

**Comment by jrudess** [comment]
That behavior has to do with ffmpeg. Once the script launches ffmpeg, if the network dies and recovers fast enough, then ffmpeg doesn't exit, and you get the small jump in recording once ffmpeg starts to get more data.

**Comment by snickerpop** [comment]
Sometimes the conversions have this tag 
Calling an asynchronous function without callback is deprecated

I am not sure what it means...

**Comment by jrudess** [comment]
I haven't tracked the root cause of that down yet - for now it's harmless, but there is a possibility of breakage on a major node.js release.   I'll hopefully have that one resolved before it becomes a problem.

**Comment by jrudess** [comment]
There are still some cases where loss of network will kill the script.

events.js:161
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at exports._errnoException (util.js:1023:11)
    at TCP.onread (net.js:572:26)

**Comment by forsomefun** [comment]
Yes, I've been getting the same error.
I didn't save it the first time but for me, it looks like for whatever reason recording was stopped at around 7 am. 

I am not sure if that is the program or my network. I am going to look online and see if their is a way to test the uptime. see if there are any drops in connection.

**Comment by jrudess** [comment]
Release 0.3.0 upgrades MFCAuto library, and the new version has fixes for network stability.   More testing is needed, but I believe this will help.

**Comment by pusspounder** [comment]
JRUDESS, maybe rename the issue to something like `Crash when internet connectivity changes'?!

This is still a problem as of today. Losing internet connection will cause the script to fail with the following:

```
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at exports._errnoException (util.js:1018:11)
    at TCP.onread (net.js:672:26)
```

BTW this is not a *"harmless"* issue IMHO. I'll try increasing the model scan interval.

**Comment by forsomefun** [comment]
@pusspounder 
Your program is broken as well. However the update from the main sstativa program works.

**Comment by jrudess** [comment]
I consider this low to medium priority, because it is easy to automatically restart the process when it happens.

https://github.com/foreverjs/forever

**Comment by forsomefun** [comment]
is that foreverjs being implemented? Or how would someone install it with your program?

**Comment by jrudess** [comment]
It is a separate program you can use to launch mfc_cb.node.  If the mfc_cb.node process ever ends, it will restart it.

**Comment by pusspounder** [comment]
*I consider this low to medium priority, because it is easy to automatically restart the process when it happens.* Hmmm...

Several issues with `forever` (this is on Windows):

Using `forever` shows multiple FFMPEG.exe windows (one for each stream I guess).
Sometimes streams aren't terminated properly (you have multiple FFMPEG.exe windows and file sizes are 0 KB, restarting seems to fix this (`forever stop`, `forever start`)

**Comment by jrudess** [comment]
Will try out `forever` from a windows VM.  I have never used this script from windows, so am uncertain of specific issues that might be faced there.

With v0.3.2, there were some problems with ffmpeg processes getting incorrectly launched twice for a model, leaving one of those processes running forever.  I believe those problems are solved with v0.3.3.  At least, I've had the script up in `forever` for 4 days now with no left over ffmpeg processes, but there could still be some corner cases not squashed.   But I don't think that problem will likely be related to `forever` itself.

If you are able to reproduce hung ffmpeg processes with v0.3.3, please open a new issue and paste some debug logs (debug: true in `config.yml`).  The printouts usually contain enough info now to track down where the race conditions might be occurring. 

The underlying problem with the script dying when network connections are lost appears to be a problem in a library used by mfc_cb.  Unfortunately, the node stack trace is completely useless to identify which library is at fault.  If anyone has better suggestions on how to root-cause the offending code, that would help a lot.

**Comment by jaller94** [comment]
Hi, just wanna let you people know that I've got the very similar stacktrace with Node 7.5.0.
This seems not to be related to your script in any way. I'm using the following:
 * official http.createServer()
 * nodejs-websocket.createServer() from NPM.

I randomly found this issue when inputing my stacktrace into Google.
```
events.js:161
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at exports._errnoException (util.js:1023:11)
    at TCP.onread (net.js:572:26)
```

**Comment by jaller94** [comment]
Updated to Node 7.10.0 and it's here again with (once again) different line numbers.
I'll tell you if I find a solution.

```
events.js:163
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at exports._errnoException (util.js:1050:11)
    at TCP.onread (net.js:582:26)
```

**Comment by jaller94** [comment]
Your project only seems to use the websocket client. Still, I would like to share my findings with you.
It stopped failing when I added an EventListener on error to each new connection.

```JavaScript
const ws = require("nodejs-websocket");

const server = ws.createServer(function (conn) {
  // without this bind the server fails sometimes
  conn.on('error', function () {
    console.log('error');
  });
}).listen(8001);

**Comment by jrudess** [comment]
Thanks for all the info @jaller94.  I wasn't even using the websocket library anymore, and removed it from the package.json.

What doesn't make sense to me, however, is that I was not even loading the nodejs-websocket library with a require() call anymore.  But I've tested the script out by shutting my router off and on, and have not been able to replicate the exception after removing the websocket library completely.

This issue should be resolved in v0.4.   If anyone still sees it happen, please update the sighting.
